includes/class-newspack-listings-products.php

```diff
@@ -0,0 +1,245 @@
+final class Newspack_Listings_Products {
+
+	/**
+	 * The option name for the product ID.
+	 */
+	const NEWSPACK_LISTINGS_PRODUCT_OPTION = 'newspack_listings_product_id';
+
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var Newspack_Listings_Products
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Is WooCommerce active? If not, we can't use any of its functionality.
+	 *
+	 * @var $wc_is_active
+	 */
+	protected static $wc_is_active = false;
+
+	/**
+	 * Main Newspack_Listings_Products instance.
+	 * Ensures only one instance of Newspack_Listings_Products is loaded or can be loaded.
+	 *
+	 * @return Newspack_Listings_Products - Main instance.
+	 */
+	public static function instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		add_action( 'init', [ __CLASS__, 'init' ] );
+
+		// When product settings are updated, make sure to update the corresponding WooCommerce products as well.
+		add_action( 'update_option', [ __CLASS__, 'update_products' ], 10, 3 );
+	}
+
+	/**
+	 * After WP init.
+	 */
+	public static function init() {
+		// Check whether WooCommerce is active and available.
+		if ( class_exists( 'WooCommerce' ) ) {
+			self::$wc_is_active = true;
+			self::create_products();
+		}
+	}
+
+	/**
+	 * Create the WooCommerce products for self-serve listings.
+	 */
+	public static function create_products() {
+		if ( ! self::$wc_is_active ) {
+			return false;
+		}
+
+		$product_id = get_option( self::NEWSPACK_LISTINGS_PRODUCT_OPTION, false );
+		if ( ! $product_id ) {
+			$settings     = Settings::get_settings();
+			$product_name = __( 'Self-Serve Listings', 'newspack-listings' );
+
+			// Parent product.
+			$parent_product = new \WC_Product_Grouped();
+			$parent_product->set_name( $product_name );
+			$parent_product->set_catalog_visibility( 'hidden' );
+			$parent_product->set_virtual( true );
+			$parent_product->set_downloadable( true );
+			$parent_product->set_sold_individually( true );
+
+			// Single listing product.
+			$single_product = new \WC_Product_Simple();
+			/* translators: %s: Product name */
+			$single_product->set_name( sprintf( __( '%s: Single Listing', 'newspack-listings' ), $product_name ) );
+			$single_product->set_regular_price( $settings['newspack_listings_single_price'] );
+			$single_product->update_meta_data( '_newspack_listings_product_slug', 'newspack_listings_single_price' );
+			$single_product->set_virtual( true );
+			$single_product->set_downloadable( true );
+			$single_product->set_catalog_visibility( 'hidden' );
+			$single_product->set_sold_individually( true );
+			$single_product->save();
+
+			// Single "featured" listing add-on.
+			$featured_addon_single = new \WC_Product_Simple();
+			/* translators: %s: Product name */
+			$featured_addon_single->set_name( sprintf( __( '%s: “Featured” Listing Add-On', 'newspack-listings' ), $product_name ) );
```

Comment:
Suggestion: "Add-On" sounds odd in this context, WDYT about "Upgrade"?

Reply:
Agreed—"add-on" came from a publisher feature request, but I like "upgrade" better.